### PR TITLE
Leader based replication - data structure updates to cache metadata response

### PR DESCRIPTION
**Pull request by Arun-LinkedIn**
Data structure changes to cache the metadata response information exchanged during a replication cycle.

### CODE REVIEW

File: ambry-replication/src/main/java/com/github/ambry/replication/RemoteReplicaInfo.java [hunk 1]
```diff
@@ -62,6 +62,10 @@
   private long reEnableReplicationTime = 0;
   private ReplicaThread replicaThread;
 
+  // Metadata response information (contains missing blobs, remote token, local lag from remote, etc.) of previous replication cycle.
```
Comment (zzmao) [inline ambry-replication/src/main/java/com/github/ambry/replication/RemoteReplicaInfo.java:65]: Minor: the comment looks like super long. There is a line in IDE, usually we don't code or comment after that line.

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java [hunk 2]
```diff
@@ -1103,24 +1118,37 @@ ReplicationMetrics getReplicationMetrics() {
...
+      this.missingStoreMessages = null;
+      this.remoteToken = null;
+      this.localLagFromRemoteInBytes = -1;
       this.serverErrorCode = errorCode;
     }
+
+    /**
+     * Utility method to extract store keys from messages
+     */
+    public static Set<StoreKey> getStoreKeysFromMessages(Set<MessageInfo> messageInfos) {
```
Comment (zzmao) [inline ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:1145]: don't need to be public.

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java [hunk 3]
```diff
@@ -1103,24 +1118,37 @@ ReplicationMetrics getReplicationMetrics() {
   }
 
   static class ExchangeMetadataResponse {
```
Comment (zzmao) [inline ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:1120]: Can we make it an individual class?

File: ambry-replication/src/main/java/com/github/ambry/replication/RemoteReplicaInfo.java [hunk 4]
```diff
@@ -198,6 +204,24 @@ public boolean equals(Object obj) {
...
+  synchronized ReplicaThread.ExchangeMetadataResponse getExchangeMetadataResponse() {
+    return exchangeMetadataResponse;
+  }
+
+  /**
+   * Stores/caches the meta data response received in the previous replication cycle.
+   * Replica threads calls this method to store the metadata response for standby replicas.
+   * @param exchangeMetadataResponse contains the meta data response information (missing keys, token info, local lag from remote, etc.).
+   */
+  synchronized void setExchangeMetadataResponse(ReplicaThread.ExchangeMetadataResponse exchangeMetadataResponse) {
```
Comment (zzmao) [inline ambry-replication/src/main/java/com/github/ambry/replication/RemoteReplicaInfo.java:221]: I need a little bit context here: when is the case to `setExchangeMetadataResponse `?

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java [hunk 5]
```diff
@@ -480,6 +494,11 @@ public void replicate() {
               // reason is replicaSyncUpManager may also throw exception and add one more exchangeMetadataResponse
               // associated with same RemoteReplicaInfo.
               exchangeMetadataResponseList.add(exchangeMetadataResponse);
+
+              // Also, store the metadata exchange received from remote replica in the RemoteReplicaInfo.
```
Comment (jsjtzyy) [inline ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:498]: minor: `store the metadata exchange response received from ...`

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java [hunk 6]
```diff
@@ -1103,24 +1131,37 @@ ReplicationMetrics getReplicationMetrics() {
...
+      this.missingStoreMessages = null;
+      this.remoteToken = null;
+      this.localLagFromRemoteInBytes = -1;
       this.serverErrorCode = errorCode;
     }
+
+    /**
+     * Utility method to extract store keys from messages
+     */
+    static Set<StoreKey> getStoreKeysFromMessages(Set<MessageInfo> messageInfos) {
```
Comment (jsjtzyy) [inline ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:1158]: According to the usage of this method, looks like it doesn't need to be a static utility method. I think a non-static method like `getMissingStoreKeys` would be more appropriate. Then the code is something like this:
```
return missingStoreMessages == null ? Collections.emptySet() : missingStoreMessages.stream().map(MessageInfo::getStoreKey).collect(Collectors.toSet());
```

File: ambry-replication/src/main/java/com/github/ambry/replication/RemoteReplicaInfo.java [hunk 7]
```diff
@@ -62,6 +63,9 @@
   private long reEnableReplicationTime = 0;
   private ReplicaThread replicaThread;
 
+  // Metadata response information received for this replica in the most recent replication cycle.
+  private ReplicaThread.ExchangeMetadataResponse exchangeMetadataResponse;
```
Comment (jsjtzyy) [inline ambry-replication/src/main/java/com/github/ambry/replication/RemoteReplicaInfo.java:67]: Add a little more comments here regarding what we need from `ExchangeMetadataResponse` for leader-based replication.

File: ambry-replication/src/main/java/com/github/ambry/replication/RemoteReplicaInfo.java (hunk 4 above)
Comment (jsjtzyy) [reply ambry-replication/src/main/java/com/github/ambry/replication/RemoteReplicaInfo.java:220]: I am trying to understand why this method should be `synchronized` ?

File: ambry-replication/src/main/java/com/github/ambry/replication/RemoteReplicaInfo.java [hunk 8]
```diff
@@ -198,6 +204,23 @@ public boolean equals(Object obj) {
     return false;
   }
 
+  /**
+   * Get the meta data response information received for this replica in the most recent replication cycle.
+   * @return exchangeMetadataResponse contains the meta data response (missing keys, token info, local lag from remote, etc.).
+   */
+  synchronized ReplicaThread.ExchangeMetadataResponse getExchangeMetadataResponse() {
+    return exchangeMetadataResponse;
```
Comment (jsjtzyy) [inline ambry-replication/src/main/java/com/github/ambry/replication/RemoteReplicaInfo.java:212]: When you working on future PR, keep in mind that `exchangeMetadataResponse` will be updated by multiple threads.

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java [hunk 9]
```diff
@@ -494,19 +514,25 @@ public void replicate() {
                 // Must have just been stopped, just skip it and move on.
                 logger.info("Local store not started for remote replica: {}", remoteReplicaInfo.getReplicaId());
                 exchangeMetadataResponseList.add(new ExchangeMetadataResponse(ServerErrorCode.Temporarily_Disabled));
+                remoteReplicaInfo.setExchangeMetadataResponse(
```
Comment (jsjtzyy) [inline ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:517]: One minor comment that should be considered in future PR: if it metadata exchange fails (i.e. due to replica unavailable) within current replication cycle, we may allow standby to send metadata request again in next cycles.

**Pull request merged**